lib/msf/base/sessions/command_shell.rb

```diff
@@ -179,28 +166,288 @@ def cmd_sessions(*args)
+    print_line
+    print_line("Downloads remote files to the local machine.")
+    print_line("This command does not support to download a FOLDER yet")
+    print_line
+  end
+
+  def cmd_download(*args)
+    if args.length != 2
+      # no argumnets, just print help message
+      return cmd_download_help
+    end
+
+    src = args[0]
+    dst = args[1]
+
+    # Check if src exists
+    if !file_exists(src)
+      print_error("The target file does not exists")
+      return
+    end
+
+    # Get file content
+    print_status("Download #{src} => #{dst}")
+    content = shell_command("cat #{src}")
+
+    # Write file to local machine
+    file = File.open(dst, "wb")
+    file.write(content)
+    file.close
+    print_good("Done")
+  end
+
+  def cmd_upload_help
+    print_line("Usage: upload [src] [dst]")
+    print_line
+    print_line("Uploads load file to the victim machine.")
+    print_line("This command does not support to upload a FOLDER yet")
+    print_line
+  end
+
+  def cmd_upload(*args)
+    if args.length != 2
+      # no argumnets, just print help message
+      return cmd_upload_help
+    end
+
+    src = args[0]
+    dst = args[1]
+
+    # Check target file exists on the target machine
+    if file_exists(dst)
+      print_warning("The file <#{dst}> already exists on the target machine")
+      if prompt_yesno("Overwrite the target file <#{dst}>?")
+        # Create an empty file on the target machine
+        # Notice here does not check the permission of the target file (folder)
+        # So if you generate a reverse shell with out redirection the STDERR
+        # you will not realise that the current user does not have permission to write to the target file
+        # IMPORTANT:
+        #   assume(the current have the write access on the target file)
+        #   if (the current user can not write on the target file) && (stderr did not redirected)
+        #     No error reporting, you must check the file created or not manually
+        result = shell_command_token("cat /dev/null > #{dst}")
+        if !result.empty?
+          print_error("Create new file on the target machine failed. (#{result})")
+          return
+        end
+        print_good("Create new file on the target machine succeed")
+      else
+        return
+      end
+    end
+
+    buffer_size = 0x100
+
+    begin
+      # Open local file
+      src_fd = open src
+      # Get local file size
+      src_size = File.size(src)
+      # Calc how many time to append to the remote file
+      times = src_size / buffer_size + (src_size % buffer_size == 0 ? 0 : 1)
+      print_status("File <#{src}> size: #{src_size}, need #{times} times writes to upload")
+      # Start transfer
+
+      for i in 1..times do
+        print_status("Uploading (#{i * buffer_size}/#{src_size})")
+        chunk = src_fd.read(buffer_size)
+        chunk_repr = repr(chunk)
+        result = shell_command_token("echo -ne '#{chunk_repr}' >> #{dst}")
```

Comment:
`echo -n` is not POSIX. You should be able to do the same with `printf` command and only POSIX compliant options. https://www.unix.com/man-page/posix/1posix/printf/

Reply:
https://github.com/rapid7/metasploit-framework/blob/55d8360705702f5cbb389eaad0205da4dd292eda/lib/msf/core/post/file.rb#L442-L572